src/rascal/math/kvec_generator.hh

```diff
@@ -0,0 +1,229 @@
+      double detM{};
+
+     public:
+      /** Constructor:
+       * As part of the constructor, the precompute() function is
+       * automatically called which computes the k-vectors for the
+       * provided parameters.
+       * @param cutoff Cutoff radius of reciprocal space ball.
+       *     As a guideline, if the smearing parameter of the density
+       *     is given by \f$sigma\f$, use \f$pi/sigma\f$ as cutoff.
+       * @param basisvectors 3x3 matrix containing the three basis
+       *     vectors a1, a2, a3 of the real space unit cell, stored as:
+       *     a1 = basisvectors[0], a2=basisvectors[1], etc.
+       *     These are then automatically converted to the corresponding
+       *     reciprocal cell vectors b1, b2, b3 by the formulae:
+       *     \f{equation}{b_1 = 2\pi \frac{a_1 \times a_2}
+       *                 {a_1 \cdot (a_2 \times a_3)} \f}
+       *     and similarly for the periodic permutations.
+       *     It is also possible to provide the basis vectors of the
+       *     reciprocal lattice directly by using the option below.
+       * @param is_reciprocal_cell false by default.
+       *     If set to true, it indicates that the cell provided
+       *     in the variable basisvectors is already the reciprocal cell.
+       *     Thus, the formula above for converting the a_i's to the b_i's
+       *     is not used, and the provided cell is used directly.
+       * @param need_origin false by default.
+       *     If set to true, the list of found k-vectors will also
+       *     contain the origin k=(0,0,0), which will then be stored
+       *     as the very first element in the matrix of found vectors.
+       */
+      explicit Kvectors(double cutoff, Eigen::Matrix3d basisvectors,
+                        bool is_reciprocal_cell = false,
+                        bool need_origin = false) {
+        // Set internal variables apart from cell
+        this->kcutoff = cutoff;
+        this->include_origin = need_origin;
+
+        /* Store provided basis vectors into the internal variable basisvecs.
+         * We need to distinguish whether the provided vectors are already
+         * reciprocal cell vectors or need to be converted first
+         */
+        if (not is_reciprocal_cell) {  // real space cell is given
+          // Create reciprocal cell from real space cell
+          Eigen::Matrix3d tcell = basisvectors.transpose();
+          this->basisvecs = 2.0 * M_PI * tcell.inverse();
+        } else {  // cell of reciprocal space is already given
+          this->basisvecs = basisvectors;
+        }
+
+        // Run the precompute function described below
+        this->precompute();
+      }  // end constructor
+
+      /**
+       * Using the provided parameters in the constructor, find
+       * all the k-vectors within the cutoff radius and store them
+       * as well as derived quantities that can be returned to the user.
+       */
+      void precompute();
```

Comment:
this function is never supposed to be called by the user right? Then it should be at least made private, or even better removed and only be part of the constructor.